File: src/main/java/com/example/demo/filters/LoggingFilter.java
package com.example.demo.filters;

import java.io.IOException;

import org.springframework.stereotype.Component;
import org.springframework.web.filter.OncePerRequestFilter;

import jakarta.servlet.FilterChain;
import jakarta.servlet.ServletException;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;


@Component
public class LoggingFilter extends OncePerRequestFilter {

    @Override
    protected void doFilterInternal(HttpServletRequest request, HttpServletResponse response, FilterChain filterChain) throws ServletException, IOException {
        long startTime = System.currentTimeMillis();
        filterChain.doFilter(request, response);
        long duration = System.currentTimeMillis() - startTime;

        String logMessage = String.format("request method: %s,request URI: %s,response status: %d,request processing time: %d ms",
                request.getMethod(), request.getRequestURI(), response.getStatus(), duration);
 
        logger.debug(logMessage);
    }
}

File: src/main/kotlin/com/example/demo/filters/LoggingFilter.kt
package com.example.demo.filters

import jakarta.servlet.FilterChain
import jakarta.servlet.ServletException
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.springframework.lang.NonNull
import org.springframework.stereotype.Component
import org.springframework.web.filter.OncePerRequestFilter
import java.io.IOException


@Component
class LoggingFilter : OncePerRequestFilter() {

    @Throws(ServletException::class, IOException::class)
    override fun doFilterInternal(@NonNull request: HttpServletRequest,
                                  @NonNull response: HttpServletResponse,
                                  filterChain: FilterChain) {

        val startTime = System.currentTimeMillis()
        filterChain.doFilter(request, response)
        val duration = System.currentTimeMillis() - startTime

        val logMessage = String.format(
            "request method: %s,request URI: %s,response status: %d,request processing time: %d ms",
            request.method, request.requestURI, response.status, duration
        )

        logger.debug(logMessage)
    }
}